src/app/ipv/middleware.js

```diff
@@ -319,213 +313,237 @@ async function handleUnexpectedPage(req, res, pageId) {
+  res.render(getIpvPageTemplatePath(PAGES.PYI_ATTEMPT_RECOVERY), {
+    csrfToken: req.csrfToken(),
+  });
+}
+
+async function updateJourneyState(req, res, next) {
+  try {
+    const currentPageId = req.params.pageId;
+    const action = req.params.action;
+
+    if (action && isValidIpvPage(currentPageId)) {
+      await handleJourneyResponse(req, res, action, currentPageId);
+    } else {
+      render404(res);
     }
-  },
-  handleJourneyPage: async (req, res, next) => {
-    try {
-      const { pageId } = req.params;
-      const { context } = req?.session || "";
-
-      // handles page id validation first
-      if (!isValidPage(pageId)) {
-        res.status(HTTP_STATUS_CODES.NOT_FOUND);
-        return res.render(getTemplatePath("errors", "page-not-found"));
-      }
+  } catch (error) {
+    next(error);
+  }
+}
 
-      if (!req.session?.ipvSessionId) {
-        logError(
-          req,
-          {
-            pageId: pageId,
-            expectedPage: req.session.currentPage,
-          },
-          "req.ipvSessionId is null",
-        );
-
-        req.session.currentPage = PAGES.PYI_TECHNICAL;
-        return res.render(getIpvPageTemplatePath(req.session.currentPage), {
-          context: "unrecoverable",
-        });
-      } else if (pageId === PAGES.PYI_TIMEOUT_UNRECOVERABLE) {
-        req.session.currentPage = PAGES.PYI_TIMEOUT_UNRECOVERABLE;
-        return res.render(getIpvPageTemplatePath(req.session.currentPage));
-      } else if (req.session.currentPage !== pageId) {
-        return await handleUnexpectedPage(req, res, pageId);
-      }
+async function handleJourneyPage(req, res, next) {
+  return handleJourneyPageInternal(req, res, next);
+}
 
-      const renderOptions = {
-        pageId,
-        csrfToken: req.csrfToken(),
-        context,
-      };
+async function handleJourneyPageInternal(
+  req,
+  res,
+  next,
+  pageErrorState = undefined,
+) {
+  try {
+    const { pageId } = req.params;
+    const { context } = req?.session || "";
 
-      if (pageRequiresUserDetails(pageId)) {
-        renderOptions.userDetails = await fetchUserDetails(req);
-      } else if (pageId === PAGES.PYI_TRIAGE_DESKTOP_DOWNLOAD_APP) {
-        const qrCodeUrl = appDownloadHelper.getAppStoreRedirectUrl(
-          parseContextAsPhoneType(context),
-        );
-        renderOptions.qrCode =
-          await qrCodeHelper.generateQrCodeImageData(qrCodeUrl);
-      } else if (pageId === PAGES.PYI_TRIAGE_MOBILE_DOWNLOAD_APP) {
-        renderOptions.appDownloadUrl = appDownloadHelper.getAppStoreRedirectUrl(
-          parseContextAsPhoneType(context),
-        );
-      } else if (req.query?.errorState !== undefined) {
-        renderOptions.pageErrorState = req.query.errorState;
-      } else if (req.session.currentPageStatusCode !== undefined) {
-        res.status(req.session.currentPageStatusCode);
-      }
+    // handles page id validation first
+    if (!isValidIpvPage(pageId)) {
+      render404(res);
+      return;
```

Comment:
Is there a difference between this and `return render404(res);`? because I can see in a few places we still do `return res.render` and I know @AmritSidhu has seen some issues around this too

Reply:
I don't think there's a difference - `render404()` returns void. If I see `return render404(res)` then I assume that there is a useful return value being passed along which I find misleading.
However, I think we probably do need to do `return next()` when that comes up because we don't know if `next()` is an async function (that returns Promise<void> instead of just void), so it's a bit of a mess :(

Reply:
Functionally there's no difference, but some people do it by convention to avoid mistakes where you accidentally try and render twice.

e.g. we had a bug at one point along the lines of:
```
if (someCondition) {
  render(somePage);
}
render(someDefaultPage);
```
if you always return, then you can't (easily) double-render... but the flipside is that as Dan says, it's a bit misleading because it's not actually returning anything.

Reply:
(I don't have a strong preference, but we should probably be consistent)

Reply:
It's probably safer to do `return render404()` everywhere.
There are places where the IntelliJ linter complains if you do
```
render404();
return;
```
But if you delete the `return` and then someone else adds more code in the future they could introduce a bug.

Reply:
Hmm, actually, ignore what I said about `return next();` `next()` is a wrapper provided by express rather than directly being the next middleware function.
Although for a similar reason to `return res.render();` `return next()` might be a good idea anyway.